links-tests/Test/Xrefcheck/FtpLinks.hs

```diff
@@ -0,0 +1,79 @@
+-- SPDX-FileCopyrightText: 2021 Serokell <https://serokell.io>
+--
+-- SPDX-License-Identifier: MPL-2.0
+
+module Test.Xrefcheck.FtpLinks
+  ( FtpHostOpt(..)
+  , ftpOptions
+  , test_FtpLinks
+  ) where
+
+import Data.Tagged (Tagged, untag)
+import Options.Applicative (help, long, strOption)
+import Test.Tasty (TestTree, askOption, testGroup)
+import Test.Tasty.HUnit (assertBool, assertFailure, testCase, (@?=))
+import Test.Tasty.Options as Tasty (IsOption (..), OptionDescription (Option), safeRead)
+
+import Xrefcheck.Config (Config (cVerification), VerifyConfig, defConfig)
+import Xrefcheck.Core (Flavor (GitHub))
+import Xrefcheck.Verify
+  (VerifyError (..), VerifyResult (VerifyResult), checkExternalResource, verifyErrors)
+
+import Test.Orphans ()
+
+-- | A list with all the options needed to configure FTP links tests.
+ftpOptions :: [OptionDescription]
+ftpOptions =
+  [ Tasty.Option (Proxy @FtpHostOpt)
+  ]
+
+-- | Option specifying FTP host.
+newtype FtpHostOpt = FtpHostOpt Text
+  deriving (Show, Eq)
+
+instance IsOption FtpHostOpt where
+  defaultValue = FtpHostOpt "ftp://localhost"
+  optionName = "ftp-host"
+  optionHelp = "[Test.Xrefcheck.FtpLinks] FTP host without trailing slash"
+  parseValue v = FtpHostOpt <$> safeRead v
+  optionCLParser = FtpHostOpt <$> strOption
+    (  long (untag (optionName :: Tagged FtpHostOpt String))
+    <> help (untag (optionHelp :: Tagged FtpHostOpt String))
+    )
+
+
+config :: VerifyConfig
+config = cVerification $ defConfig GitHub
+
+test_FtpLinks :: TestTree
+test_FtpLinks = askOption $ \(FtpHostOpt host) -> do
+  let notRaisedMsg = "expected exceptions was not raised"
+
+  testGroup "Ftp links handler"
+    [ testCase "handles correct link to file" $ do
+        let link = host <> "/pub/file_exists.txt"
+        result <- checkExternalResource config link
+        result @?= VerifyResult []
```

Comment:
Somehow tests pass locally, though I cannot figure out how is this possible taking into account that I have no `pub` directory or something like that. Is it some ftp magic?

Reply:
hm...

I have this output for `cabal run -- links-tests --ftp-host "ftp://statgen.ncsu.edu"`
```
Ftp links handler
      handles correct link to file:                FAIL (2.02s)
        links-tests/Test/Xrefcheck/FtpLinks.hs:56:
        expected: VerifyResult []
         but got: VerifyResult [FileDoesNotExist "pub/file_exists.txt"]
      handles empty link (host only):              OK (1.70s)
      handles correct link to non empty directory: OK (1.71s)
      handles correct link to empty directory:     FAIL (1.68s)
        links-tests/Test/Xrefcheck/FtpLinks.hs:71:
        expected: VerifyResult []
         but got: VerifyResult [FileDoesNotExist "empty"]
      throws exception when file not found:        OK (1.86s)

2 out of 5 tests failed (8.97s)
```
this output is expected by me (at this server there are no `/pub/file_exists.txt` and `/empty/`

Reply:
oh, I've got it))

https://github.com/serokell/xrefcheck/blob/79b4e106f0a6d14dd8e4715fac3bbc47644e72ee/src/Xrefcheck/Verify.hs#L295-L295

Reply:
Meeh, right :smile:

Reply:
I've added config option for localhost links

Reply:
Good! Left a comment there.